### PR TITLE
Feature/transaction history

### PR DESCRIPTION
**Pull request by drEfrati**
Payments History Feature

### CODE REVIEW

**Review by mrsegev — APPROVED** [review]
Nice, just make the changes we talked about + update Readme file to reflect the new API.

File: kin-core/src/main/java/kin/core/AccountInfoRetriever.java [hunk 1]
```diff
@@ -2,27 +2,40 @@
...
     private final Server server;
     private final KinAsset kinAsset;
+    private BlockchainEvents blockchainEvents;
 
     AccountInfoRetriever(Server server, KinAsset kinAsset) {
         this.server = server;
         this.kinAsset = kinAsset;
     }
 
+    void setBlockChainEvents(BlockchainEvents blockchainEvents) {
```
Comment (yosriz) [inline kin-core/src/main/java/kin/core/AccountInfoRetriever.java]: To be consistent with the current design - every dependency should be "injected" directly into the ctor (and then omit the null checks)

File: kin-core/src/main/java/kin/core/BlockchainEvents.java [hunk 2]
```diff
@@ -135,27 +135,35 @@ public void onEvent(TransactionResponse transactionResponse) {
...
     private void extractPaymentsFromTransaction(TransactionResponse transactionResponse,
-        EventListener<PaymentInfo> listener) {
+                                                EventListener<PaymentInfo> listener) {
+        PaymentInfo paymentInfo = getPaymentInfo(transactionResponse);
+        if (paymentInfo != null) {
+            listener.onEvent(paymentInfo);
+        }
+    }
+
+    PaymentInfo getPaymentInfo(TransactionResponse transactionResponse) {
```
Comment (yosriz) [inline kin-core/src/main/java/kin/core/BlockchainEvents.java]: extract seems like a more appropriate name

File: kin-core/src/main/java/kin/core/AccountInfoRetriever.java [hunk 3]
```diff
@@ -2,27 +2,40 @@
...
 import org.stellar.sdk.responses.AccountResponse;
 import org.stellar.sdk.responses.HttpResponseException;
+import org.stellar.sdk.responses.Page;
+import org.stellar.sdk.responses.TransactionResponse;
 
 class AccountInfoRetriever {
 
     private final Server server;
     private final KinAsset kinAsset;
+    private BlockchainEvents blockchainEvents;
```
Comment (yosriz) [inline kin-core/src/main/java/kin/core/AccountInfoRetriever.java]: can you extract the `TransactionResponse` -> `PaymentInfo` logic to a common class? it's not entirely clean to have ` AccountInfoRetriever` depends on unrelated `BlockchainEvents` class

File: kin-core/src/main/java/kin/core/AccountInfoRetriever.java [hunk 4]
```diff
@@ -62,6 +75,74 @@ Balance getBalance(@NonNull String accountId) throws OperationFailedException {
...
+        if (blockchainEvents != null) {
+            payments = getPaymentsHistory(transactionBuilder, requestParams.getAccountId());
+        }
+        return payments;
+    }
+
+    private List<PaymentInfo> getPaymentsHistory(TransactionsRequestBuilder transactions, @NonNull String accountId) throws OperationFailedException {
+        List<PaymentInfo> payments = new ArrayList<>();
+        try {
+            Page<TransactionResponse> execute = transactions
```
Comment (yosriz) [inline kin-core/src/main/java/kin/core/AccountInfoRetriever.java]: execute -> response

File: kin-core/src/main/java/kin/core/AccountInfoRetriever.java [hunk 5]
```diff
@@ -62,6 +75,74 @@ Balance getBalance(@NonNull String accountId) throws OperationFailedException {
...
+                    .forAccount(KeyPair.fromAccountId(accountId))
+                    .execute();
+            if (execute == null || execute.getRecords() == null) {
+                throw new AccountNotActivatedException(accountId);
```
Comment (yosriz) [inline kin-core/src/main/java/kin/core/AccountInfoRetriever.java]: why it's `AccountNotActivatedException` in this case?
what does null response mean? I would expect general exception (data cannot be retrieved) like in getBalance/sendTransaction

File: kin-core/src/main/java/kin/core/AccountInfoRetriever.java [hunk 6]
```diff
@@ -62,6 +75,74 @@ Balance getBalance(@NonNull String accountId) throws OperationFailedException {
...
+                }
+            }
+        } catch (HttpResponseException httpError) {
+            if (httpError.getStatusCode() == 404) {
+                throw new AccountNotFoundException(accountId);
+            } else {
+                throw new OperationFailedException(httpError);
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (yosriz) [inline kin-core/src/main/java/kin/core/AccountInfoRetriever.java]: rethrow this as `OperationFailedException`

File: kin-core/src/main/java/kin/core/BlockchainEvents.java [hunk 7]
```diff
@@ -135,27 +135,35 @@ public void onEvent(TransactionResponse transactionResponse) {
...
+                                new TransactionIdImpl(transactionResponse.getHash()),
+                                extractHashTextIfAny(transactionResponse)
                         );
-                        listener.onEvent(paymentInfo);
                     }
                 }
             }
-
         }
+        // return a new payment info object if possible, otherwise return null.
```
Comment (yosriz) [inline kin-core/src/main/java/kin/core/BlockchainEvents.java]: seems a redundant comment

File: kin-core/src/main/java/kin/core/KinAccount.java [hunk 8]
```diff
@@ -104,6 +106,63 @@ TransactionId sendTransactionSync(@NonNull String publicAddress, @NonNull BigDec
     @NonNull
     Balance getBalanceSync() throws OperationFailedException;
 
+    /**
+     * Create {@link Request} for getting all the transactions payments history for the current account.
```
Comment (yosriz) [inline kin-core/src/main/java/kin/core/KinAccount.java]: payment transactions

File: kin-core/src/main/java/kin/core/KinAccount.java [hunk 9]
```diff
@@ -104,6 +106,63 @@ TransactionId sendTransactionSync(@NonNull String publicAddress, @NonNull BigDec
...
+     * <p> See {@link KinAccount#getPaymentsHistorySync()} for possibles errors</p>
+     *
+     * @return {@code Request<List<PaymentInfo>> } PaymentInfo - the payment information for a specific payment operation in a transaction
```
Comment (yosriz) [inline kin-core/src/main/java/kin/core/KinAccount.java:113]: List of

File: kin-core/src/main/java/kin/core/KinAccount.java [hunk 10]
```diff
@@ -104,6 +106,63 @@ TransactionId sendTransactionSync(@NonNull String publicAddress, @NonNull BigDec
...
+     */
+    @NonNull
+    Request<List<PaymentInfo>> getPaymentsHistory();
+
+    /**
+     * For method description see {@link KinAccount#getPaymentsHistory()}
+     * @param requestParams an optional parameters to the request, such as accountId, cursor, limit, and order.
+     * Create {@link Request} for getting all the transactions payments history for a given account.
```
Comment (yosriz) [inline kin-core/src/main/java/kin/core/KinAccount.java]: I don't think we should expose the possibility of giving an account id, as it doesn't match the current interfaces - `KinAccount` is for a specific address and it doesn't appropriate to give an option here to query other accounts.

File: kin-core/src/main/java/kin/core/KinAccount.java [hunk 11]
```diff
@@ -104,6 +106,63 @@ TransactionId sendTransactionSync(@NonNull String publicAddress, @NonNull BigDec
...
+     *                        If no account has been given to the requestParams then it will be for the current account.
+     * <p> See {@link KinAccount#getPaymentsHistorySync()} for possibles errors</p>
+     *
+     * <p><b>Note:</b>     requestParams can have the next members:</p>
+     * <p><b>accountId</b> is optional, a string, if not given then using the current account if exist. It is represents the ID of an account.</p>
+     * <p><b>cursor:</b>   is optional, a paging token, specifying where to start returning records from. for example 12884905984.</p>
```
Comment (yosriz) [inline kin-core/src/main/java/kin/core/KinAccount.java]: I'm not sure how token can be used here? what actually it is?
We should give the option to get this token from every payment o.w. it is not useful currently

File: kin-core/src/main/java/kin/core/KinAccount.java [hunk 12]
```diff
@@ -104,6 +106,63 @@ TransactionId sendTransactionSync(@NonNull String publicAddress, @NonNull BigDec
...
+     * <p><b>order:</b>    is optional, an Order, currently the default is "asc".	It is represents the order in which to return rows, “asc” or “desc”.</p>
+     * <p><b>limit:</b>    is optional, a number, currently the default is 10. It is represents the maximum number of records to return.</p>
+     *
+     * @return {@code Request<List<PaymentInfo>> } PaymentInfo - the payment information for a specific payment operation in a transaction
+     */
+    @NonNull
+    Request<List<PaymentInfo>> getPaymentsHistory(PaymentsHistoryRequestParams requestParams);
+
+    /**
+     * Get the list of the transactions payments history for the current account.
```
Comment (yosriz) [inline kin-core/src/main/java/kin/core/KinAccount.java]: limit doc - maybe something like: default limit is 10, use `getPaymentsHistorySync(PaymentsHistoryRequestParams)` for providing a differnt limit

File: kin-core/src/main/java/kin/core/KinAccountImpl.java [hunk 13]
```diff
@@ -57,6 +62,25 @@ public Balance getBalanceSync() throws OperationFailedException {
...
+    @Override
+    public List<PaymentInfo> getPaymentsHistorySync() throws OperationFailedException {
+        checkValidAccount();
+        return accountInfoRetriever.getPaymentsHistory(account.getAccountId());
+    }
+
+    @NonNull
+    @Override
+    public List<PaymentInfo> getPaymentsHistorySync(PaymentsHistoryRequestParams requestParams) throws OperationFailedException {
+        String accountId = requestParams.getAccountId();
```
Comment (yosriz) [inline kin-core/src/main/java/kin/core/KinAccountImpl.java]: see comment above regarding specifying account id

File: kin-core/src/main/java/kin/core/PaymentsHistoryRequestParams.java [hunk 14]
```diff
@@ -0,0 +1,129 @@
+package kin.core;
+
+
+import android.os.Parcel;
+import android.os.Parcelable;
+
+/**
+ * Class which build optional params for the transaction payment history request.
```
Comment (yosriz) [inline kin-core/src/main/java/kin/core/PaymentsHistoryRequestParams.java]: Represents optional params for payment history request, see <add link to the method>

File: kin-core/src/main/java/kin/core/PaymentsHistoryRequestParams.java [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    private PaymentsHistoryRequestParams(PaymentsHistoryRequestParamsBuilder builder) {
+        this.accountId = builder.accountId;
+        this.token = builder.token;
+        this.limit = builder.limit;
+        this.order = builder.order;
+    }
+
+    protected PaymentsHistoryRequestParams(Parcel in) {
```
Comment (yosriz) [inline kin-core/src/main/java/kin/core/PaymentsHistoryRequestParams.java]: why does it need to be parcelable?

File: kin-core/src/main/java/kin/core/PaymentsHistoryRequestParams.java [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+    public int getLimit() {
+        return limit;
+    }
+
+    public Order getOrder() {
+        return order;
+    }
+
+
+    public static class PaymentsHistoryRequestParamsBuilder {
```
Comment (yosriz) [inline kin-core/src/main/java/kin/core/PaymentsHistoryRequestParams.java]: just Builder (`new PaymentsHistoryRequestParams.Builder`)

File: kin-core/src/main/java/kin/core/PaymentsHistoryRequestParams.java [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+
+        private String accountId;           //optional
+        private String token;               //optional
+        private int limit;                  //optional
+        private Order order;                //optional
+
+        public PaymentsHistoryRequestParamsBuilder account(String accountId) {
```
Comment (yosriz) [inline kin-core/src/main/java/kin/core/PaymentsHistoryRequestParams.java]: add "set" prefix to be consistent with other builders

Comment (yosriz) [inline kin-core/src/main/java/kin/core/PaymentsHistoryRequestParams.java]: add docs for setters

**Review by yosriz — CHANGES_REQUESTED** [review]
:+1:
The most important comments are regarding the public APIs (the ability to change account id, the token)
Regarding tests, please see codecov report:
https://codecov.io/gh/kinecosystem/kin-core-android/commit/2534f9f47b3c99f2df00420ae8f1ddc117b28be4 
I'm not sure why codecov didn't put a comment on the PR like it does always (maybe you need to sign in?)

File: kin-core/src/main/java/kin/core/KinAccount.java [hunk 18]
```diff
@@ -104,6 +106,57 @@ TransactionId sendTransactionSync(@NonNull String publicAddress, @NonNull BigDec
     @NonNull
     Balance getBalanceSync() throws OperationFailedException;
 
+    /**
+     * Create {@link Request} for getting list of payments history for the current account.
+     * The default is currently the last 10 payments, please use use {@link KinAccount#getPaymentsHistory((PaymentsHistoryRequestParams)} for providing a different limit.
```
Comment (yosriz) [inline kin-core/src/main/java/kin/core/KinAccount.java:111]: "currently" is redundant, a suggestion: "The default limit is 10 payments".

Comment (yosriz) [inline kin-core/src/main/java/kin/core/KinAccount.java:111]: redundant "use" (please use use)

File: kin-core/src/main/java/kin/core/KinAccount.java (hunk 9 above)
Comment (yosriz) [reply kin-core/src/main/java/kin/core/KinAccount.java:113]: the return value doc should specify it is the list of payments of this account.

File: kin-core/src/main/java/kin/core/PaymentInfoHelper.java [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+import org.stellar.sdk.Operation;
+import org.stellar.sdk.PaymentOperation;
+import org.stellar.sdk.responses.TransactionResponse;
+import java.math.BigDecimal;
+import java.util.List;
+
+/**
+ * Helper class which handle the extraction of Payment Info from a transaction.
+ */
+public class PaymentInfoHelper {
```
Comment (yosriz) [inline kin-core/src/main/java/kin/core/PaymentInfoHelper.java:14]: should internal class with all the methods internal as well

File: kin-core/src/main/java/kin/core/PaymentsHistoryRequestParams.java [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+        return order;
+    }
+
+
+    public static class Builder {
+
+        private int limit;                  //optional
+        private Order order;                //optional
+
+//        /**
```
Comment (yosriz) [inline kin-core/src/main/java/kin/core/PaymentsHistoryRequestParams.java:51]: dead code comment - should be deleted

**Review by yosriz — COMMENTED** [review]
@drEfrati I've added small comments for small tweaks, other than that it's all good

**Pull request closed without merging**